inference/engine.go

```diff
@@ -26,38 +26,43 @@ import (
 	"golang.org/x/tools/go/analysis"
 )
 
+// conflictHandler defines the interface that handles the conflicts encountered during inference.
+// This makes the inference engine independent of the diagnostic generation logic.
+type conflictHandler interface {
+	AddSingleAssertionConflict(trigger annotation.FullTrigger)
+	AddOverconstraintConflict(nilExplanation, nonnilExplanation ExplainedBool)
+}
+
 // Engine is the structure responsible for running the inference: it contains methods to run
 // various tasks for the inference and stores an internal map that can be obtained by calling
-// Engine.InferredMapWithDiagnostics.
+// Engine.InferredMap.
 type Engine struct {
 	pass *analysis.Pass
 	// inferredMap is the internal inferred map that the engine writes to, it is initialized on the
 	// construction of the engine and populated by the "Observe*" methods of the engine. Users
-	// should use the Engine.InferredMapWithDiagnostics() method to obtain the current inferred map.
+	// should use the Engine.InferredMap() method to obtain the current inferred map.
 	inferredMap *InferredMap
-	// conflicts stores conflicts encountered during the observations. It will be
-	// converted to diagnostics and returned along with the current inferred map whenever users
-	// request it.
-	conflicts *ConflictList
+	// diagnosticEngine receives all encountered conflicts during inference and eventually
+	// generates proper diagnostics from those conflicts.
+	diagnosticEngine conflictHandler
 	// controlledTriggersBySite stores the set of controlled triggers for each site if the site
 	// controls any triggers. This field is for internal use in the struct only and should not be
 	// accessed elsewhere.
 	controlledTriggersBySite map[primitiveSite]map[annotation.FullTrigger]bool
 }
 
 // NewEngine constructs an inference engine that is ready to run inference.
-func NewEngine(pass *analysis.Pass) *Engine {
+func NewEngine(pass *analysis.Pass, diagnosticEngine conflictHandler) *Engine {
 	return &Engine{
-		pass:        pass,
-		inferredMap: newInferredMap(),
-		conflicts:   &ConflictList{},
+		pass:             pass,
+		inferredMap:      newInferredMap(),
+		diagnosticEngine: diagnosticEngine,
 	}
 }
 
-// InferredMapWithDiagnostics returns the current inferred annotation map and a slice of diagnostics
-// generated during inference.
-func (e *Engine) InferredMapWithDiagnostics() (*InferredMap, []analysis.Diagnostic) {
-	return e.inferredMap, e.conflicts.Diagnostics()
+// InferredMap returns the current inferred annotation map.
+func (e *Engine) InferredMap() *InferredMap {
```

Comment:
This separates diagnostic generation logic from inference, callers should call `.Diagnostics()` on the diagnostic engine instead.